web3/contract.py

```diff
@@ -1442,6 +1450,96 @@ def factory(cls, class_name, **kwargs):
         return PropertyCheckingFactory(class_name, (cls,), kwargs)
 
 
+class ContractCaller:
+    """
+    An alternative Contract API.
+
+    This call:
+
+    > contract.caller({'from': eth.accounts[1], 'gas': 100000, ...}).add(2, 3)
+    is equivalent to this call in the classic contract:
+    > contract.functions.add(2, 3).call({'from': eth.accounts[1], 'gas': 100000, ...})
+
+    Other options for invoking this class include:
+
+    > contract.caller.add(2, 3)
+
+    or
+
+    > contract.caller().add(2, 3)
+
+    or
+
+    > contract.caller(transaction_dict={'from': eth.accounts[1], 'gas': 100000, ...}).add(2, 3)
+    """
+    def __init__(self,
+                 abi,
+                 web3,
+                 address,
+                 transaction_dict=None,
+                 block_identifier='latest'):
+        self.web3 = web3
+        self.address = address
+        self.abi = abi
+        self._functions = None
+
+        if abi:
+            if transaction_dict is None:
+                transaction_dict = {}
+
+            self._functions = filter_by_type('function', self.abi)
+            for func in self._functions:
+                fn = ContractFunction.factory(
+                    func['name'],
+                    web3=self.web3,
+                    contract_abi=self.abi,
+                    address=self.address,
+                    function_identifier=func['name'])
+
+                block_id = parse_block_identifier(self.web3, block_identifier)
+                caller_method = partial(self.call_function,
+                                        fn,
+                                        transaction_dict=transaction_dict,
+                                        block_identifier=block_id)
+
+                setattr(self, func['name'], caller_method)
+
+    def __getattr__(self, function_name):
+        if self.abi is None:
+            raise NoABIFound(
+                "There is no ABI found for this contract.",
+            )
+        if not self._functions or len(self._functions) == 0:
+            raise NoABIFunctionsFound(
+                "The ABI for this contract contains no function definitions. ",
+                "Are you sure you provided the correct contract ABI?"
+            )
+        elif function_name not in self._functions:
+            functions_available = ', '.join([fn['name'] for fn in self._functions])
+            raise MismatchedABI(
+                "The function '{}' was not found in this contract's ABI. ".format(function_name),
+                "Here is a list of all of the function names found: {}".format(functions_available),
+                "Did you mean to call one of those functions?"
```

Comment:
This will result in a message like 

> `"Here is a list of all of the function names found: fn_a, fn_bDid you mean to call one of those functions?"

Notice the lack of space before the `Did`.

Reply:
Oops, nice catch!